rfcs/FINEGRAINED_NOTIFICATIONS.md

```diff
@@ -0,0 +1,122 @@
+
+This is interesting for a number of use cases:
+
+1) I want to animate or refresh a RecyclerView when elements are added, moved, modified or deleted in a `RealmResults`.
+2) I want to be notified about the result of an query changing so I can refresh my UI.
+3) I want my changelisteners to _only_ trigger when there is a relevant change.
+
+While you in some use cases probably want to compare old and new data, we have not yet found a compelling enough use case to support this for collections.
+Single object notifications will have this functionality.
+
+## Current situation
+
+- Right now Realm Java changelisteners are "table-based", which means that a change to any element in a table will trigger all changelisteners registered on that table, even though the element wasn't included in the query. This results in a lot of un-needed change events being triggered.
+
+- Single object change listeners work the same way, which is even worse.
+
+- Changelisteners are not supported on RealmList
+
+- Cocoa has support for fine-grained collection changes.
+
+- `Realm` / `RealmResults` / `RealmObject` all have methods like `addChangeListener/removeChangeListener/removeAllListeners`.
+
+
+## Detailed design
+
+Add these new interfaces and class:
+
+```
+// New generic interface for generic changes (= our current changelisteners)
+// All relevant classes should implement this: Realm, RealmResults, RealmList, RealmObject, DynamicRealm, DynamicRealmObject
+// TODO: Is Observable the best name?
+public interface Observable<T> {
+    void addChangeListener(RealmChangeListener<T> listener);
+    void removeChangeListener(RealmChangeListener<T> listener);
+    void removeAllChangeListeners(); // Current name is removeChangeListeners? IMO this name is better, but do we want the penalty of a rename?
+}
+
+// New interface for collections. Add another changelistener for fine-grained notifications
+public interface ObservableCollection<T, L> extends Observable<T> {
+    void addChangeListener(L listener);
+    void removeChangeListener(L listener);
+}
+
+// New listener interface for fine-grained callback. Changes are encapsulated in the OrderedCollectionChange object
+// Using OrderedCollectionChange to prepare for change events from unordered data structures.
+public interface OrderedRealmCollectionChangeListener<T> {
+    void onChange(T collection, OrderedCollectionChange changes);
+}
+
+// Class for describing changes
+// Primary purpose is to make RecyclerView animations easy, but should also be general purpose
+// RecyclerView uses the `get*Ranges()` methods.
+public class OrderedCollectionChange {
+    public long[] getDeletions() { return null; }
+    public long[] getInsertertions() { return null; }
+    public long[] getChanges()  { return null; }
```

Comment:
I went with "modifications" mostly because I found that the word "changes" was getting too overloaded when writing the docs. Using the same word in the class name and one of the property names makes it hard to word things.